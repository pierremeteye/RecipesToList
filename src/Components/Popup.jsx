import React, { Component } from 'react';

export default class Popup extends Component {
    constructor(props) {
        super(props);
        this.state = {
            ingredients: [],
            quantity: [],
            list: ''
        };
    }

    addToShoppingList(el, i) {
        this.props.data[this.props.getValue].ingredients.map((el, i)  => {
            let quantity = (el.quantity === 0 ? '' : ' / ' + Math.round(el.quantity)) + '' + (el.quantity === 0 ? '' : el.unit)
            this.state.ingredients.push(el.name)
            this.state.quantity.push(Math.round(el.quantity))
        })
        this.props.getNumber(this.state.quantity);
        this.props.getIngredients(this.state.ingredients);

        let arr = this.state.ingredients,
            list = {}
        ;

        for (var i = 0; i < arr.length; i++) {
            if (list[this.state.ingredients[i]] === undefined) {
                list[this.state.ingredients[i]] = this.state.quantity[i];
            } else {
                list[this.state.ingredients[i]] += this.state.quantity[i];   
            }
        }
        this.setState({
            list: list
        })        
        console.log(this.state.list);
    }

    render() {
        if (this.props.getValue === null) {
            return '';
        } else {
            return  <div>
                        <h2 className="col-12 f-l center">{this.props.data[this.props.getValue].title} / <span className="green">{this.props.data[this.props.getValue].servings} persons</span></h2>
                        <div className="col-12 f-l p-t-45">
                            <div className="col-6 f-l">
                                <h3 className="col-12 f-l p-b-10">Ingredients List</h3>
                                <ul className="col-12 f-l noListStyle p-b-10">
                                    {
                                        this.props.data[this.props.getValue].ingredients.map((el, i)  => {
                                            return <li className="uppercaseFirstLetter" key={i}>{el.name}<span className="red">{el.quantity === 0 ? '' : ' / ' + el.quantity} {el.quantity === 0 ? '' : el.unit}</span></li>;
                                        })
                                    }
                                </ul>
                            </div>
                            <div className="col-6 f-l">
                                <img className="col-8 offset-col-2 f-l" src={'images/' + this.props.data[this.props.getValue].image_name} alt={this.props.data[this.props.getValue].title}/>
                            </div>
                        </div>
                        <div className="col-12 f-l">
                            <h3 className="p-t-10 p-b-10">Instructions</h3>
                            <p>{this.props.data[this.props.getValue].instructions}</p>
                        </div>
                        <div className="col-12 f-l p-t-30 p-b-30 center">
                            <button onClick={this.addToShoppingList.bind(this)} className="pointer">Add this recipe to my shopping list</button>
                        </div>
                    </div>
                                  
        }
    }
}