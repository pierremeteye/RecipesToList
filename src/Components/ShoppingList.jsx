import React, {Component} from "react";

export default class ShoppingList extends Component {

	render(){
        if (this.props.data) {
			let arrQuantity = Object.values(this.props.list),
				arrIngredients = Object.keys(this.props.list),
                ingredients = arrIngredients.map((ingredient, i) => {
                    return <tr key={ingredient + '-' + i}><td>{ingredient}</td></tr>;
                }),
                quantity = arrQuantity.map((quantity, i) => {
                    return <tr key={quantity + '-' + i}><td>{quantity}</td></tr>;
                })

			;

			return 	<div className="ShoppingList col-12">
			            <div className="col-6 f-l">
		                    <div className="p-30">
		                        <h3 className="col-12 f-l p-b-10">Ingredients List</h3>
								<div>
									<div className="both"></div>
									<table className="f-l">
										<tbody>
											<tr>
												<th>Ingredients</th>
											</tr>
											{ingredients}
										</tbody>
									</table>
									<table className="f-l">
										<tbody>
											<tr>
												<th>Quantity</th>
											</tr>
											{quantity}
										</tbody>
									</table>
								</div>
		                    </div>
		                </div>
		            </div>
		}
	}
}