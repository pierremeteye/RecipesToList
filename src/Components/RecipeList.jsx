import React, { Component } from 'react';

import Popup from "./Popup";

class RecipeList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            value: null,
            isHidden: false,
            ingredients: null
        };
    }
    getIngredientsPopup(dataFromPopup) {
        this.props.getIngredient(dataFromPopup)
    }
    getNumber(getNumber){
        this.props.getNumber(getNumber)
    }
    showPopup(i) {
        this.setState(prevState => ({
            value: i,
            isHidden: true
        }));
        let lis = this.props.initialIngredients[i].map((el, i)  => {
            return <li className="uppercaseFirstLetter" key={i}>{el.name}<span className="red">{el.quantity === 0 ? '' : ' / ' + Math.round(el.quantity)} {el.quantity === 0 ? '' : el.unit}</span></li>;
        });
        this.setState({ingredients: lis})
    }
    closePopup(){
        this.setState(prevState => ({
            isHidden: false
        }));
    }
    render() {
		var recipes = this.props.data.map((recipe, i) => {
			return  <div key={i} className="col-12 f-l relative m-t-10">
        				<div className="col-8 f-l">
                            <p className="col-8 f-l">{recipe.title}</p>
                            <button className="f-l pointer" onClick={this.showPopup.bind(this, i)}>Voir la recette</button>
                        </div>
                        <div className="col-4 f-l">
                            <img className="col-4 f-l" src={'images/' + recipe.image_name} alt={recipe.title}/>
                        </div>
        			</div>;
		});
        return (
            <div className="RecipeList col-12">
	            <div className="col-6 f-l">
                    <div className="p-30">
                    <h3 className="col-12 f-l p-b-10">Ingredients List</h3>
                        {recipes}
                        <div className={this.state.isHidden === true ? 'popup' : 'popup hide' }>
                            <div className="col-6 bgWhite fullCentered p-20 p-t-0">
                                <p className="pointer" onClick={this.closePopup.bind(this)}>
                                    <i className="zmdi zmdi-close fs-50 f-r p-10 p-r-0"></i>
                                </p>
                                <Popup 
                                    data={this.props.data} 
                                    ingredients={this.state.ingredients} 
                                    getValue={this.state.value} 
                                    getIngredients={this.getIngredientsPopup.bind(this)}
                                    getNumber={this.getNumber.bind(this)}
                                />
                            </div>
                        </div>
                        <div className="both"></div>
                    </div>
                </div>
            </div>
        );
    }
}

export default RecipeList;