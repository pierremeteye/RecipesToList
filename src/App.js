import React, { Component } from 'react';

// Components
import RecipeList from './Components/RecipeList';
import ShoppingList from './Components/ShoppingList';

class App extends Component {

    constructor(props) {
        super(props);
        this.state = {
            error: null,
            isLoaded: false,
            items: [],
            initialIngredients: [],
            ingredientsFromPopup: [],
            quantity: [],
            list: {}
        };
    }
    getIngredientsList(dataFromList){
        this.setState({list: dataFromList})
    }
    getNumber(getNumber){
        this.setState({quantity: getNumber})
    }
    componentDidMount() {
        fetch("recipes.json")
        .then(res => res.json())
        .then(
            (result) => {
                let getItemsArr = [];
                let newGetItemsArr = [];
                result.map((el, i) => {
                   return getItemsArr.push(el.ingredients);
                });
                getItemsArr.map((el, i) => {
                   return newGetItemsArr.push(el);
                });
                this.setState({
                    isLoaded: true,
                    items: result,
                    initialIngredients: getItemsArr
                });
            },
            (error) => {
                this.setState({
                    isLoaded: true,
                    error
                });
            }
            );
            // var getItems = this.state.items.map(function(index, elem) {
            //     console.log(index);
            //     return index.ingredients;
            // })
            // this.setState({initialIngredients: getItems})

    }
    render() {
        const { error, isLoaded } = this.state;
        if (error) {
            return <div>Error: {error.message}</div>;
        } else if (!isLoaded) {
            return <div>Loading...</div>;
        } else {
            return (
                <div id="main">
                    <RecipeList 
                        data={this.state.items} 
                        initialIngredients={this.state.initialIngredients} 
                        shoppinglist={this.state.ingredientsFromPopup} 
                        getIngredient={this.getIngredientsList.bind(this)}
                        getNumber={this.getNumber.bind(this)}
                       
                    />
                    <ShoppingList 
                        data={this.state.items} 
                        list={this.state.list}
                        ingredients={this.state.ingredientsFromPopup}
                        quantity={this.state.quantity}
                    />
                </div>
            );
        }
    }
}

export default App;