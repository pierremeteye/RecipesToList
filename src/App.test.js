import React from "react";
import Enzyme, { shallow, render, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { createSerializer } from "enzyme-to-json";
import sinon from "sinon";

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));

Enzyme.configure({ adapter: new Adapter() });

// Components import
import App from './App.js';
import Popup from './Components/Popup.jsx';
import RecipeList from './Components/RecipeList.jsx';
import ShoppingList from './Components/ShoppingList.jsx';

it("renders correctly App Component", () => {
	const AppTest = shallow(
		<App />
		);
	expect(AppTest);
  // expect(AppTest).toMatchSnapshot();
});

it("renders correctly Popup Component", () => {
	const dataPopup = {
		title: 'test',
		arr: ['test'],
		value: 0
	}
	const PopupTest = shallow(
		<Popup 
		data={dataPopup.title} 
		getValue={dataPopup.value}
		initialIngredients={dataPopup.arr}
		/>
		);
	expect(PopupTest);
  // expect(PopupTest).toMatchSnapshot();
});

it("renders correctly RecipeList Component", () => {
	const dataRecipeList = {
		arr: ['test']
	}
	const RecipeListTest = shallow(
		<RecipeList 
		data={dataRecipeList.arr} 
		/>
		);
	expect(RecipeListTest);
  // expect(RecipeListTest).toMatchSnapshot();
});

it("renders correctly ShoppingList Component", () => {
	const ShoppingListTest = shallow(
		<ShoppingList />
		);
	expect(ShoppingListTest);
  // expect(ShoppingListTest).toMatchSnapshot();
});