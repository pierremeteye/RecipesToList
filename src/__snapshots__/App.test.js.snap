// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly App Component 1`] = `
<div>
  Loading...
</div>
`;

exports[`renders correctly Popup Component 1`] = `
<div>
  <h2
    className="col-12 f-l center"
  >
     / 
    <span
      className="green"
    >
       persons
    </span>
  </h2>
  <div
    className="col-12 f-l p-t-45"
  >
    <div
      className="col-6 f-l"
    >
      <h3
        className="col-12 f-l p-b-10"
      >
        Ingredients List
      </h3>
      <ul
        className="col-12 f-l noListStyle p-b-10"
      />
    </div>
    <div
      className="col-6 f-l"
    >
      <img
        className="col-8 offset-col-2 f-l"
        src="images/undefined"
      />
    </div>
  </div>
  <div
    className="col-12 f-l"
  >
    <h3
      className="p-t-10 p-b-10"
    >
      Instructions
    </h3>
    <p />
  </div>
  <div
    className="col-12 f-l p-t-30 p-b-30 center"
  >
    <button
      className="pointer"
      onClick={[Function]}
    >
      Add this recipe to my shopping list
    </button>
  </div>
</div>
`;

exports[`renders correctly RecipeList Component 1`] = `
<div
  className="RecipeList col-12"
>
  <div
    className="col-6 f-l"
  >
    <div
      className="p-30"
    >
      <h3
        className="col-12 f-l p-b-10"
      >
        Ingredients List
      </h3>
      <div
        className="col-12 f-l relative m-t-10"
        key="0"
      >
        <div
          className="col-8 f-l"
        >
          <p
            className="col-8 f-l"
          />
          <button
            className="f-l pointer"
            onClick={[Function]}
          >
            Voir la recette
          </button>
        </div>
        <div
          className="col-4 f-l"
        >
          <img
            className="col-4 f-l"
            src="images/undefined"
          />
        </div>
      </div>
      <div
        className="popup hide"
      >
        <div
          className="col-6 bgWhite fullCentered p-20 p-t-0"
        >
          <p
            className="pointer"
            onClick={[Function]}
          >
            <i
              className="zmdi zmdi-close fs-50 f-r p-10 p-r-0"
            />
          </p>
          <Popup
            data={
              Array [
                "test",
              ]
            }
            getIngredients={[Function]}
            getNumber={[Function]}
            getValue={null}
            ingredients={null}
          />
        </div>
      </div>
      <div
        className="both"
      />
    </div>
  </div>
</div>
`;

exports[`renders correctly ShoppingList Component 1`] = `""`;
